src/java/org/apache/cassandra/schema/SchemaManager.java

```diff
@@ -0,0 +1,617 @@
+     */
+    public Optional<Function> findFunction(FunctionName name, List<AbstractType<?>> argTypes)
+    {
+        if (!name.hasKeyspace())
+            throw new IllegalArgumentException(String.format("Function name must be fully quallified: got %s", name));
+
+        KeyspaceMetadata ksm = getKeyspaceMetadata(name.keyspace);
+        return ksm == null
+               ? Optional.empty()
+               : ksm.functions.find(name, argTypes);
+    }
+
+    public void applyReceivedSchemaMutationsOrThrow(InetAddressAndPort from, Collection<Mutation> payload)
+    {
+        updateHandler.asGossipAwareTrackerOrThrow("Received schema push request from " + from).applyReceivedSchemaMutations(from, payload);
+    }
+
+    public Collection<Mutation> prepareRequestedSchemaMutationsOrThrow(InetAddressAndPort from)
+    {
+        return updateHandler.asGossipAwareTrackerOrThrow("Received schema pull request from " + from).prepareRequestedSchemaMutations(from);
+    }
+
+    public Schema schema()
+    {
+        return updateHandler.schema();
+    }
+
+    void applyChangesLocally(KeyspacesDiff diff)
+    {
+        diff.dropped.forEach(this::dropKeyspace);
+        diff.created.forEach(this::createKeyspace);
+        diff.altered.forEach(this::alterKeyspace);
+    }
+
+    private void alterKeyspace(KeyspaceDiff delta)
+    {
+        SchemaDiagnostics.keyspaceAltering(schema(), delta);
+
+        // update keyspace metadata
+        Keyspace keyspace = getKeyspaceInstance(delta.before.name);
+        assert keyspace != null;
+
+        assert delta.before.name.equals(delta.after.name);
+
+        // drop tables and views
+        delta.tables.dropped.forEach(t -> dropTable(keyspace, t));
+        delta.views.dropped.forEach(v -> dropView(keyspace, v));
+
+        // add tables and views
+        delta.tables.created.forEach(t -> createTable(keyspace, t));
+        delta.views.created.forEach(v -> createView(keyspace, v));
+
+        // update tables and views
+        delta.tables.altered.forEach(diff -> alterTable(keyspace, diff.after));
+        delta.views.altered.forEach(diff -> alterView(keyspace, diff.after));
+
+        // update keyspace metadata in keyspace instance
+        keyspace.setMetadata(delta.after);
+
+        // deal with all added, and altered views
+        keyspace.viewManager.reload(true);
+
+
+        schemaChangeNotifier.notifyKeyspaceAltered(delta);
+    }
+
+    private void createKeyspace(KeyspaceMetadata keyspace)
+    {
+        SchemaDiagnostics.keyspaceCreating(schema(), keyspace);
+        Keyspace.open(keyspace.name);
+        schemaChangeNotifier.notifyKeyspaceCreated(keyspace);
+    }
+
+    private void dropKeyspace(KeyspaceMetadata keyspace)
+    {
+        SchemaDiagnostics.keyspaceDroping(schema(), keyspace);
+
+        // remove the keyspace from the static instances.
+        removeKeyspaceInstance(keyspace.name, ks -> {
+            keyspace.views.forEach(v -> dropView(ks, v));
```

Comment:
I went through the original, DSE and STAR-790 and this is how it looks for dropping keyspace:

Original scenario:
1. apply changes to system_schema
2. drop views/tables
3. remove keyspace instance
4. unload CFs
5. remove keyspace metadata
6. remove table refs
7. barrier
8. update version

DSE:
1. apply changes to system_schema
2. update the current schema
3. update table refs
4. drop tables/views
5. remove keyspace instance
6. unload CFs
7. barrier
8. update version

STAR-790:
1. apply changes to system_schema
2. update the current schema
3. update table refs
4. remove keyspace instance
5. drop views/tables
6. unload CFs
7. barrier
8. update version

The differences are between 2-6. Let me explain a bit:
- removing keyspace instace is just dereferencing the instance of `Keyspace` so that no one gets it - In STAR-790, what is different to DSE and original, keyspace instance is dropped in the first place, before dropping tables and views, and before unloading CFs. I justify this by the fact I don't want to return the keyspace instance which is being modified (dropping tables and views, and unloading CFs are made on keyspace instance). Once the removal is started, the keyspace instance is no longer available.
- removing keyspace metadata - it is updating the runtime schema by removing the keyspace metadata; in STAR-790 and DSE this is a separate entity which can be updated externally by pulling changes from somewhere. As such, changes in table refs and keyspace instances are secondary to the changes in runtime schema, so they should go afterwards. Though, this deserved longer discussion probably. I would postpone it for another ticket - for now, we have dse-like order in this regards